metadata-ingestion/src/datahub/ingestion/source/powerbi.py

```diff
@@ -0,0 +1,1382 @@
+
+
+class PowerBiDashboardSourceConfig(PowerBiAPIConfig):
+    platform_name: str = "powerbi"
+    platform_urn: str = "urn:li:dataPlatform:powerbi"
+    dashboard_pattern: AllowDenyPattern = AllowDenyPattern.allow_all()
+    chart_pattern: AllowDenyPattern = AllowDenyPattern.allow_all()
+
+
+class PowerBiAPI:
+    # API endpoints of PowerBi to fetch dashboards, tiles, datasets
+    API_ENDPOINTS = {
+        Constant.DASHBOARD_LIST: "{POWERBI_BASE_URL}/{WORKSPACE_ID}/dashboards",
+        Constant.ENTITY_USER_LIST: "{POWERBI_ADMIN_BASE_URL}/{ENTITY}/{ENTITY_ID}/users",
+        Constant.TILE_LIST: "{POWERBI_BASE_URL}/{WORKSPACE_ID}/dashboards/{DASHBOARD_ID}/tiles",
+        Constant.DATASET_GET: "{POWERBI_BASE_URL}/{WORKSPACE_ID}/datasets/{DATASET_ID}",
+        Constant.DATASOURCE_GET: "{POWERBI_BASE_URL}/{WORKSPACE_ID}/datasets/{DATASET_ID}/datasources",
+        Constant.REPORT_GET: "{POWERBI_BASE_URL}/{WORKSPACE_ID}/reports/{REPORT_ID}",
+        Constant.SCAN_GET: "{POWERBI_ADMIN_BASE_URL}/workspaces/scanStatus/{SCAN_ID}",
+        Constant.SCAN_RESULT_GET: "{POWERBI_ADMIN_BASE_URL}/workspaces/scanResult/{SCAN_ID}",
+        Constant.SCAN_CREATE: "{POWERBI_ADMIN_BASE_URL}/workspaces/getInfo",
+    }
+
+    @dataclass
+    class Workspace:
+        """
+        PowerBi Workspace
+        """
+
+        id: str
+        name: str
+        state: str
+        dashboards: List[Any]
+        datasets: Dict
+
+    @dataclass
+    class DataSource:
+        """
+        PowerBi
+        """
+
+        @dataclass
+        class MetaData:
+            """
+            MetaData about DataSource
+            """
+
+            is_relational: Boolean
+
+        id: str
+        type: str
+        database: Optional[str]
+        server: Optional[str]
+        metadata: Any
+
+        def __members(self):
+            return (self.id,)
+
+        def __eq__(self, instance):
+            return (
+                isinstance(instance, PowerBiAPI.DataSource)
+                and self.__members() == instance.__members()
+            )
+
+        def __hash__(self):
+            return hash(self.__members())
+
+    # dataclasses for PowerBi Dashboard
+    @dataclass
+    class Dataset:
+        @dataclass
+        class Table:
+            name: str
+            schema_name: str
+
+        id: str
+        name: str
+        webUrl: Optional[str]
+        workspace_id: str
+        datasource: Any
+        # Table in datasets
+        tables: List[Any]
+
+        def get_urn_part(self):
+            return "datasets.{}".format(self.id)
+
+        def __members(self):
+            return (self.id,)
+
+        def __eq__(self, instance):
+            return (
+                isinstance(instance, PowerBiAPI.Dataset)
+                and self.__members() == instance.__members()
+            )
+
+        def __hash__(self):
+            return hash(self.__members())
+
+    @dataclass
+    class Report:
```

Comment:
didn't see this in the Concept Mapping section. What is Report being mapped to?

Reply:
Tile is either created from Report, Dataset, or Visualization. If Tile is created from Report then chart.externalUrl is set to Report.webUrl